### ISSUE TITLE
Enahcnement Request: Small problem with "libs": ["sqlite3"]

### ISSUE THREAD

**Issue by aldacron** [biozic/d2sqlite3#15]
While I understand the motivation to include the sqlite3 library in the project configuration, I think this is only truly useful on non-Windows systems. On those systems, it's a safe assumption that the library will be on the system path and that it will be in the format required by the linker. On Windows, those assumptions cannot be made.

Given that we have multiple linkers to deal with (OPTLINK, link, and ld), and that there is no such thing as a common path for installed development libraries, most Windows developers tend to have their own custom configurations set up. Sometimes that may involve configuring a global library path, other times it may be based on a naming convention.

Here's my situation. I'm working on a book for PACKT called "Learning D". In the chapter I'm presently finishing off, I develop a sample vibe.d app with sqlite3 as the DB and d2sqlite3 to interface with it. The book uses DMD exclusively. In order to make things as painless for the Windows-based readers as possible, I've provided three versions of the sqlite3 library that I link via a pragma(lib) statement depending on the backend being used. 

```
version(Windows) {
    // When compiling with -m64
    version(Win64)
        enum sqliteLib = "lib\\sqlite3-ms64.lib";
    // When compiling with -m32mscoff
    else version(CRuntime_Microsoft)
        enum sqliteLib = "lib\\sqlite3-ms32.lib";
    // The default, or when compiling with -m32
    else
        enum sqliteLib = "lib\\sqlite3-dm-dll.lib";

    pragma(lib, sqliteLib);
}
```
Readers on other systems can be served by the libs directive in dub.json.

When compiling with OPTLINK, the libs directive causes a warning that sqlite3.lib can't be found, but compilation succeeds and the app runs. When compiling with the -m64 flag (-ax86_64 with dub), Microsoft's link errors out when it can't find the library. I only discovered that last night as I was testing everything before I prepare to submit the chapter.

There are ways around this on my end. For example, I could set up a configuration that tells the linker to ignore sqlite3.lib on platform=windows-x86_64, but that doesn't help when the user wants to compile with the -m32mscoff option. DUB doesn't currently support that with the -a flag, but it will in the future. Also, I have no idea if it will be possible to distinguish between the two 32-bit runtimes from dub.json in the future.

Aside from this specific case, I intend to use d2sqlite3 for future projects (I love how easy it is to use). This issue is going to disrupt my workflow. I generally build libraries from source and add suffixes to the names similar to those in the example above. I won't be able to do that with sqlite3 if I intend to use the MS linker (which I do!). 

I propose that you add two configurations to the bottom of dub.json. The first, which would be the default, could include the libs directive, while the second, perhaps named something like "no-default-lib" would not. Then I, and others like me, could specify the second configuration when we need to and for everyone else it would be business as usual.

**Comment by MartinNowak** [comment]
You could ship the package with a corresponding sqlite import library or maybe even w/ static libraries.

**Comment by aldacron** [comment]
The fact that I needed to bundle the libraries for 32-bit COFF and OMF as well as 64-bit COFF is what caused me to find the issue in the first place. This issue goes beyond my source package for the book, anyway. DUB configurations for compiling on Windows should never assume a standard library name on the global library path. It is not at all unusual for developers to build libraries with names different from those provided by the default build system.

**Comment by MartinNowak** [comment]
As sqlite is public domain, it makes sense to bundle this package w/ precompiled libs, also see vibe.d.
https://github.com/rejectedsoftware/vibe.d/tree/master/lib

**Comment by biozic** [comment]
I agree. I'll look into it, but I'am pessimistic about my Windows skills...

**Comment by MartinNowak** [comment]
You should be able to simply compile the amalgamated file with dmc and cl.
I can actually do this easily using our release virtualbox images.
https://www.sqlite.org/howtocompile.html

**Comment by biozic** [comment]
I had no problems using cl to compile object files for the latest sqlite amalgamation. Everything works (see the branch called windows-thrill).

I couldn't compile with dmc: too many errors. I might have missed the right switch. So instead I tried to link to the dll I downloaded from sqlite.org. I made an import lib from the .def file with lib.exe, and then converted it with coffimplib.exe. But when I run the unittests, there's an access violation in some part of the FTS5 code. Maybe I should try the import lib @aldacron provides with his book.
